### PR TITLE
Add plots for V0M vs ntrklt

### PR DESCRIPTION
**Pull request by njacazio**
Do not merge

### CODE REVIEW

File: machine_learning_hep/data/data_prod_20200304/database_ml_parameters_D0pp_0304.yml [hunk 1]
```diff
@@ -145,9 +145,9 @@ D0pp:
       fracmerge: [0.08,0.08,0.08] #list of periods
       seedmerge: [12,12,12] #list of periods
       period: [LHC16pp,LHC17pp,LHC18pp] #list of periods
-      unmerged_tree_dir: [/data/TTree/D0DsLckAnywithJets/vAN-20200304_ROOT6-1/pp_2016_data/374_20200304-2028/merged,
-                          /data/TTree/D0DsLckAnywithJets/vAN-20200304_ROOT6-1/pp_2017_data/375_20200304-2028/merged,
-                          /data/TTree/D0DsLckAnywithJets/vAN-20200304_ROOT6-1/pp_2018_data/376_20200304-2028/merged] #list of periods
+      unmerged_tree_dir: [/tmp/TTree/D0DsLckAnywithJets/vAN-20200304_ROOT6-1/pp_2016_data/374_20200304-2028/merged,
```
Comment (benedikt-voelkel) [inline machine_learning_hep/data/data_prod_20200304/database_ml_parameters_D0pp_0304.yml:148]: We should not set the paths in the default database to a value different different from what it was before. Otherwise, people might not find their data in the future.
In general, changing paths in the databases in a PR can easily cause confusion. 
If really needed, we should make a validation database

Comment applies to many other lines

File: machine_learning_hep/do_entire_analysis.py [hunk 2]
```diff
@@ -1,5 +1,5 @@
 #############################################################################
-##  © Copyright CERN 2018. All rights not expressly granted are reserved.  ##
+##  © Copyright CERN 2018. All rights not expressly granted are reserved.  ##
```
Comment (benedikt-voelkel) [inline machine_learning_hep/do_entire_analysis.py:2]: We should not change the copyright. That makes the continuous integration fail

File: machine_learning_hep/utilities_validation.py [hunk 3]
```diff
@@ -30,6 +30,44 @@ def fillvalidationvsmult(dfevt, dfevtevtsel, df_reco):
     Create a TH3F histogram and fill it with three variables from a dataframe.
     """
     _ = len(df_reco)
+    hlist = []
+    def makenfill(nx, xl, xh, ny, yl, yh, X, Y):
```
Comment (benedikt-voelkel) [inline machine_learning_hep/utilities_validation.py:34]: I would suggest to have variable names with more meaningful names. The package became so large that it otherwise might be hard to follow what is going on

File: machine_learning_hep/utilities_validation.py [hunk 4]
```diff
@@ -30,6 +30,44 @@ def fillvalidationvsmult(dfevt, dfevtevtsel, df_reco):
...
+      hn = "h_" + X + "_vs_" + Y
```
Comment (benedikt-voelkel) [inline machine_learning_hep/utilities_validation.py:35]: Just a thing which might be helpful/fancy: Python has so called formatted strings and you can make them putting an f before the quotation marks, like
`hn = f"f_{X}_vs_{Y}"`

File: machine_learning_hep/do_entire_analysis.py (hunk 2 above)
Comment (njacazio) [reply machine_learning_hep/do_entire_analysis.py]: My bad ;)

File: machine_learning_hep/utilities_validation.py (hunk 3 above)
Comment (njacazio) [reply machine_learning_hep/utilities_validation.py:34]: I plan to use `makefill2dhist` as defined in `utilities_plot.py`

File: machine_learning_hep/utilities_validation.py (hunk 4 above)
Comment (njacazio) [reply machine_learning_hep/utilities_validation.py:35]: Indeed, thanks, this will be useful

File: machine_learning_hep/submission/default_histomass.yml [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  type: YYYY # used unless specified explicitly as do_entire_analysis -a type_ana
+  # Do each period separately including merged (true)
+  # Do only merged (false)
+  doperperiod: false
+  data:
+    histomass: true # processer: process_histomass
+    dojetsystematics: false # analyzer: jetsystematics
+  mc:
+    histomass: true # processer: process_histomass
+    efficiency: false # processer: process_efficiency, process_response, process_unfolding
```
Comment (vkucera) [inline machine_learning_hep/submission/default_histomass.yml:65]: What is the benefit of having a separate config file with only histomass enabled? If the point here is to run only the processor part of the analysis in a script, I would advocate for making a default_processor.yml file instead with efficiency enabled as well which makes more sense in terms of running different blocks of the framework separately. You can always change the flags locally to your current needs but the committed changes should be useful in a longer term.

File: machine_learning_hep/utilities.py [hunk 6]
```diff
@@ -116,34 +116,26 @@ def create_folder_struc(maindir, listpath):
...
-
-def checkdir(mydir):
+def checkdir(mydir, V=True):
     """
     Checks if folder already exist, to not overwrite by accident
     """
     exfolders = 0
     if os.path.exists(mydir):
+        if V:
+          print("Please remove this directory:")
```
Comment (vkucera) [inline machine_learning_hep/utilities.py:126]: I am not sure this is very useful. In the current version, one can copy the block of lines from the standard output and paste it directly in the command line to delete the directories. This change would add this comment between each pair of lines which would make it less smooth to execute the rm commands for all the directories.

Comment (vkucera) [reply machine_learning_hep/utilities.py:126]: Also I think one letter variable names are not allowed in the style check because they are not very meaningful most of the time.

File: machine_learning_hep/utilities.py [hunk 7]
```diff
@@ -116,34 +116,26 @@ def create_folder_struc(maindir, listpath):
...
-    Checks if list of folder already exist, to not overwrite by accident
-    """
-    exfolders = 0
-    for _, mydir in enumerate(dirlist):
-        if os.path.exists(mydir):
-            print("rm -rf ", mydir)
-            exfolders = exfolders - 1
-    return exfolders
-
-def checkdir(mydir):
```
Comment (vkucera) [inline machine_learning_hep/utilities.py:130]: Do I get it right that you just changed the order of the functions? Sorry, it is not easy to see from the diff whether there are other changes.

File: machine_learning_hep/utilities_validation.py (hunk 3 above)
Comment (vkucera) [reply machine_learning_hep/utilities_validation.py:34]: I think the comment applies more to the variables `nx, xl, xh, ny, yl, yh, X, Y`. They are indeed quite cryptic and I think they would not pass the pylint check.

File: machine_learning_hep/utilities_validation.py [hunk 8]
```diff
@@ -30,6 +30,44 @@ def fillvalidationvsmult(dfevt, dfevtevtsel, df_reco):
...
+        a = [yl]
+        while a[-1] + 0.1 < 10:
+          a.append(a[-1] + 0.1)
+        while a[-1] + 1 < 100:
+          a.append(a[-1] + 1)
+        ar = TArrayF(len(a))
+        for i,j in enumerate(a):
+          ar.SetAt(j, i)
+        h = TH2F(hn, ht, nx, xl, xh, len(a) - 1, 0, 1)
+        h.GetYaxis().Set(len(a) - 1, ar.GetArray())
```
Comment (vkucera) [inline machine_learning_hep/utilities_validation.py:48]: You don't need to create a `TArrayF` here. You can use `array('d', a)` directly in the `TH2F` constructor. You just need to import `array` like this: `from array import array`

File: machine_learning_hep/utilities_validation.py [hunk 9]
```diff
@@ -30,6 +30,44 @@ def fillvalidationvsmult(dfevt, dfevtevtsel, df_reco):
     Create a TH3F histogram and fill it with three variables from a dataframe.
```
Comment (vkucera) [inline machine_learning_hep/utilities_validation.py:30]: This doesn't look like the proper description of what the function does.

**Comment by ginnocen** [comment]
hi all, thanks for all the comments. But just to clarify, Nicolo's request was really just make to show me what he did so far. It was not meant to be merged.

**Comment by vkucera** [comment]
> hi all, thanks for all the comments. But just to clarify, Nicolo's request was really just make to show me what he did so far. It was not meant to be merged.

Ah, OK, sorry, I thought it was a regular PR.

**Comment by ginnocen** [comment]
No worry Vit, it is actually always very good you all keep an eye on the new PRs 👍

**Comment by njacazio** [comment]
Thank you for all your comments, indeed this was not intended to be merged so far.
Anyway it is useful to get a feedback, I will come with a refined PR so don't be surprised if the solutions here are temporary.
Apologies ;)

File: machine_learning_hep/utilities_validation.py (hunk 3 above)
Comment (njacazio) [reply machine_learning_hep/utilities_validation.py:34]: Yes, I fully agree, my comment related to the fact that it had to be reimplemented.

File: machine_learning_hep/submission/default_histomass.yml (hunk 5 above)
Comment (njacazio) [reply machine_learning_hep/submission/default_histomass.yml:65]: This has no sense at all, it was just for me to get a grasp on the code. It will go away together with the [WIP]

File: machine_learning_hep/utilities.py (hunk 6 above)
Comment (njacazio) [reply machine_learning_hep/utilities.py:126]: Also here, I agree, it was just to know what the code was telling

File: machine_learning_hep/utilities.py (hunk 7 above)
Comment (njacazio) [reply machine_learning_hep/utilities.py:130]: This might have been an artifact of the rebasing, I will have to check this

File: machine_learning_hep/utilities_validation.py (hunk 8 above)
Comment (njacazio) [reply machine_learning_hep/utilities_validation.py:48]: Yes, indeed, I will use the functions as defined in `utilities_plot.py`

File: machine_learning_hep/utilities_validation.py (hunk 9 above)
Comment (njacazio) [reply machine_learning_hep/utilities_validation.py:30]: I agree, nonetheless, it has nothing to do with this PR, I will fix it though

**Comment by ginnocen** [comment]
that is my fault

**Comment by njacazio** [comment]
Ok, now things should look more like what you are used to see on this repository.
I couldn't resist the need to change reuse the basic functions in `utilities_plot.py`, apologies ;)
I think this includes the changed we discussed this morning.
If you find it satisfactory we can merge this and proceed to the next implementation!

**Comment by ginnocen** [comment]
Hi @njacazio, this looks good to me, but i see you still have the WIP. Should I just remove it and merge?

File: machine_learning_hep/utilities_plot.py [hunk 10]
```diff
@@ -24,25 +24,61 @@
...
-    histo = TH2F(titlehist, titlehist, lenx, arrayx, leny, arrayy)
+def buildhisto(h_name, h_tit, arrayx, arrayy=None, arrayz=None):
+    """
+    Create a histogram of size 1D, 2D, 3D, depending on the number of arguments given
+    """
+    histo = None
+    def binning(binning_array):
+        return len(binning_array) - 1, binning_array
+    if arrayz:
+        histo = TH3F(h_name, h_tit, *binning(arrayx), *binning(arrayy), *binning(arrayz))
```
Comment (ginnocen) [inline machine_learning_hep/utilities_plot.py:55]: hi @njacazio, how is it * operation called?

**Comment by ginnocen** [comment]
@njacazio I will just merge it cause I want to use it for one check on the corrected multiplicity
and it just looks very good to me. If there is anything you want to update, please just do another PR. Thanks! It seems we got a guy that can code 👍

File: machine_learning_hep/utilities_plot.py (hunk 10 above)
Comment (vkucera) [reply machine_learning_hep/utilities_plot.py:55]: @ginnocen https://docs.python.org/3.6/tutorial/controlflow.html#unpacking-argument-lists

Comment (njacazio) [reply machine_learning_hep/utilities_plot.py:55]: Thank you @ginnocen
